firewood/src/merkle/stream.rs

```diff
@@ -88,197 +100,239 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+                                    let iter = std::iter::once((extension.chd(), prev_index));
+
+                                    branch_iter_stack.push(BranchIterator {
+                                        key_nibbles,
+                                        children_iter: Box::new(iter),
+                                    });
+                                    break;
+                                };
+
+                                match next_extension_nibble.cmp(&next_key_nibble) {
+                                    std::cmp::Ordering::Equal => (), // The nibbles match; check the next one.
+                                    std::cmp::Ordering::Less => break, // [extension]'s child is before [key]. Skip it.
+                                    std::cmp::Ordering::Greater => {
+                                        // [extension]'s child is after [key]. Visit it.
+                                        let child = merkle
+                                            .get_node(extension.chd())
+                                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                                        let Some(branch) = child.inner().as_branch() else {
+                                            return Poll::Ready(Some(Err(
+                                                api::Error::InternalError(Box::new(
+                                                    api::Error::InvalidExtensionChild,
+                                                )),
+                                            )));
+                                        };
+
+                                        branch_iter_stack.push(BranchIterator {
+                                            key_nibbles: matched_key_nibbles.clone(),
+                                            children_iter: Box::new(get_children_iter(branch)),
+                                        });
+                                        break;
+                                    }
+                                }
+
+                                matched_key_nibbles.push(next_key_nibble);
+                            }
+                        }
+                    }
                 }
 
-                self.key_state = IteratorState::Iterating { visited_node_path };
+                self.key_state = IteratorState::Iterating { branch_iter_stack };
 
                 self.poll_next(_cx)
             }
+            IteratorState::Iterating { branch_iter_stack } => {
+                while let Some(mut branch_iter) = branch_iter_stack.pop() {
+                    // [node_addr] is the next node to visit.
+                    // It's the child at index [pos] of [node_iter].
+                    let Some((node_addr, pos)) = branch_iter.children_iter.next() else {
+                        // We visited all this node's descendants.
+                        // Go back to its parent.
+                        continue;
+                    };
```

Comment:
🤔 ... `branch_iter_stack` could actually be a `FlatMap` such that it actually just outputs the children directly. 

I think... in any case, this can be for another PR.